linkerd/proxy/http/src/orig_proto.rs

```diff
@@ -76,24 +77,44 @@ where
 
         *req.version_mut() = http::Version::HTTP_2;
 
-        self.inner.call(req).map(|mut res| {
-            debug_assert_eq!(res.version(), http::Version::HTTP_2);
-            let version = if let Some(orig_proto) = res.headers_mut().remove(L5D_ORIG_PROTO) {
-                debug!("downgrading {} response: {:?}", L5D_ORIG_PROTO, orig_proto);
+        UpgradeFuture {
+            version,
+            inner: self.inner.call(req),
+        }
+    }
+}
+
+pub struct UpgradeFuture<F> {
+    version: http::Version,
+    inner: F,
+}
+
+impl<F, B> Future for UpgradeFuture<F>
+where
+    F: Future<Item = http::Response<B>>,
+{
+    type Item = F::Item;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let mut res = try_ready!(self.inner.poll());
+        let version = res
+            .headers_mut()
+            .remove(L5D_ORIG_PROTO)
+            .and_then(|orig_proto| {
```

Comment:
From above it looks like `orig_proto` could be either `HTTP/1.1` or `HTTP/1.1; absolute-form` (same for 1.0).

If the `L5D_ORIG_PROTO` header has the `..; absolute-form` ending then the `else` statement would set it back to `None`. That behavior isn't changed here, but is that an issue?

Reply:
I also noticed this, but after taking a closer look, I believe this isn't an issue. This future is the _response_ future, not the request, and "absolute form" only applies to requests. The request handling side _does_ handle the presence of `; absolute-form` in the `l5d-orig-proto` header.

It would be nice if there was a comment here so future readers don't have to hunt for this...

Reply:
Okay cool I agree it doesn't sound like an issue if "absolute form" only applies to requests. Thanks! And yes a comment on this definitely wouldn't hurt.